File: backend/app/api/favorites.py
"""收藏API"""
from fastapi import APIRouter, Depends, HTTPException, Query
from typing import Optional
from sqlalchemy.orm import Session

from sqlalchemy.orm import joinedload
from app.core.database import get_db
from app.models.favorite import Favorite
from app.models.model import Model3D
from app.models.user import User
from app.api.deps import get_current_user
from app.schemas.favorite import FavoriteItem, FavoriteModelBrief, FavoriteListResponse

router = APIRouter(prefix="/api/favorites", tags=["收藏"])


def _model_to_brief(m: Model3D) -> FavoriteModelBrief:
    cat = None
    if m.category:
        try: cat = m.category.value
        except: cat = str(m.category)
    st = None
    if m.status:
        try: st = m.status.value
        except: st = str(m.status).lower()
    return FavoriteModelBrief(
        id=m.id,
        name=m.name,
        category=cat,
        image_url=m.preview_images,
        glb_path=m.glb_path,
        price=m.base_price,
        status=st,
    )


@router.get("", response_model=FavoriteListResponse)
def get_favorites(
    sort: Optional[str] = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """获取当前用户的收藏列表"""
    q = db.query(Favorite).filter(Favorite.user_id == user.id)
    if sort == "oldest":
        order = Favorite.created_at.asc()
    else:
        order = Favorite.created_at.desc()
    total = q.count()
    items = (
        q.options(joinedload(Favorite.model_rel))
        .order_by(order)
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )
    return FavoriteListResponse(
        total=total,
        items=[
            FavoriteItem(
                id=f.id,
                model=_model_to_brief(f.model_rel),
                created_at=f.created_at,
            )
            for f in items
            if f.model_rel
        ],
    )


@router.post("/{model_id}", response_model=dict)
def add_favorite(
    model_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """收藏模型"""
    model = db.query(Model3D).filter(Model3D.id == model_id).first()
    if not model:
        raise HTTPException(status_code=404, detail="模型不存在")

    existing = db.query(Favorite).filter(
        Favorite.user_id == user.id,
        Favorite.model_id == model_id,
    ).first()
    if existing:
        return {"ok": True, "message": "已收藏"}

    f = Favorite(user_id=user.id, model_id=model_id)
    db.add(f)
    db.commit()
    return {"ok": True, "message": "收藏成功"}


@router.delete("/{model_id}", response_model=dict)
def remove_favorite(
    model_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """取消收藏"""
    f = db.query(Favorite).filter(
        Favorite.user_id == user.id,
        Favorite.model_id == model_id,
    ).first()
    if not f:
        raise HTTPException(status_code=404, detail="未收藏")
    db.delete(f)
    db.commit()
    return {"ok": True, "message": "已取消收藏"}


@router.get("/check/{model_id}", response_model=dict)
def check_favorite(
    model_id: int,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """检查是否已收藏"""
    f = db.query(Favorite).filter(
        Favorite.user_id == user.id,
        Favorite.model_id == model_id,
    ).first()
    return {"favorited": f is not None}

File: frontend/src/api/favorites.ts
// 收藏 API

const API_BASE = '/api'

function getAuthHeaders(): Record<string, string> {
  const token = localStorage.getItem('3dprint_token')
  return token ? { Authorization: `Bearer ${token}` } : {}
}

export interface FavoriteModelBrief {
  id: number
  name: string
  category?: string | null
  image_url?: string | null
  glb_path?: string | null
  price?: number | null
  status?: string | null
}

export interface FavoriteItem {
  id: number
  model: FavoriteModelBrief
  created_at: string
}

export interface FavoriteListResponse {
  total: number
  items: FavoriteItem[]
}

// 获取收藏列表
export async function getFavorites(page = 1, sort?: string): Promise<FavoriteListResponse> {
  let url = `${API_BASE}/favorites?page=${page}&page_size=20`
  if (sort) url += `&sort=${sort}`
  const res = await fetch(url, { headers: getAuthHeaders() })
  if (!res.ok) throw new Error('获取收藏列表失败')
  return res.json()
}

// 收藏模型
export async function addFavorite(modelId: number): Promise<void> {
  const res = await fetch(`${API_BASE}/favorites/${modelId}`, {
    method: 'POST',
    headers: getAuthHeaders(),
  })
  if (!res.ok) throw new Error('收藏失败')
}

// 取消收藏
export async function removeFavorite(modelId: number): Promise<void> {
  const res = await fetch(`${API_BASE}/favorites/${modelId}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  })
  if (!res.ok) throw new Error('取消收藏失败')
}

// 检查是否已收藏
export async function checkFavorite(modelId: number): Promise<boolean> {
  const res = await fetch(`${API_BASE}/favorites/check/${modelId}`, { headers: getAuthHeaders() })
  if (!res.ok) return false
  const data = await res.json()
  return data.favorited
}

// === 分享相关 ===

export interface MutualFollowInfo {
  is_mutual: boolean
  i_follow: boolean
  they_follow: boolean
}

/** 检查当前登录用户与目标用户是否互相关注 */
export async function checkMutualFollow(userId: number): Promise<MutualFollowInfo> {
  const res = await fetch(`${API_BASE}/users/${userId}/mutual-follow`, { headers: getAuthHeaders() })
  if (!res.ok) throw new Error('检查关注关系失败')
  return res.json()
}

/** 获取模型上传者的用户 ID */
export async function getModelUploader(modelId: number): Promise<number> {
  const res = await fetch(`${API_BASE}/models/${modelId}`, { headers: getAuthHeaders() })
  if (!res.ok) throw new Error('获取模型信息失败')
  const data = await res.json()
  return data.uploader_id
}
